rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,241 @@
+The main thing this does is use SQLite's [immutable mode](https://www.sqlite.org/c3ref/open.html).
+
+This is a separate feature;
+it is perfectly possible to implement the `local-overlay` without this or vice-versa.
+But for maximum usability, we want to do both.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Because the `local-overlay` store is a completely separate store implementation, the interactions with the rest of Nix are fairly minimal and well-defined.
+In particular, users of other stores and not the `local-overlay` store will not be impacted at all.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Read-only local store is delicate
+
+SQLite's immutuable mode doesn't mean "we promise not to change the database".
+It relies on the database not just being *logically* immutable (the meaning doesn't change), but *physically* immutable (no bytes of the on-disk files change).
+This is because it is forgoing synchronization altogether and thus relying that nothing can be rearranged e.g. in the middle of a query (invalidating in-progress traversals of data structures, etc.).
+This means there is no hope of, say "append only" mode where "publishers" only add new store objects to a local store, while read-only mode "subscribers" are able to deal with other store objects just fine.
+
+This is an inconvenience for rolling out new versions of the lower store, but not a show stopper.
+One solution is "multi version concurrency control" where consumers get an immutable snapshot of the lower store for the duration of each login.
+New snapshots can only be gotten when consumers log in again, and old snapshots can only be retired once every consumer viewing them logs out.
+
+## `local-overlay` lacks normal form for the database
+
+A slight drawback with the architecture is a lack of a normal form.
+A store object in the lower store may or may not have a DB entry in the `overlay-local` store.
+
+> In the store object state diagriam diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
+
+This introduces some flexibility in the system: the seem "logical" layered store can be represented in multiple different "physical" configurations.
+This isn't a problem per-se, but does mean there is a bit more complexity to consider during testing and system administration.
+
+## Deleting isn't intuitive
+
+For "deleting" lower store objects in the `local-overlay` store,
+we don't actually remove them but just remove the upper DB entry.
+This is somewhat surprising, but reflects the fact that the lower store is logically immutable (even when it isn't a `local` store opened in read-only mode).
+By deleting the upper DB entry, we are not removing the object from the `local-overlay` store, but we are still resetting it to the initial state.
+
+# Alternatives
```

Comment:
This seems like a good place to mention the `nix-store --load-db` alternative mentioned on the Super Colliding Nix Stores [discourse thread](https://discourse.nixos.org/t/super-colliding-nix-stores/28462/7). Downsides to that approach include:
- It takes 10 minutes to import the database for the 10TB store that Replit have.
- The database takes up 634MB (which is over half a user's quota).

Reply:
See what I wrote. I didn't see `nix-store --load-db` in the blog post unless I am missing something?

Reply:
Ah the *discourse* thread.

Reply:
Added something about that too.